osidb/mixins.py

```diff
@@ -241,3 +242,179 @@ def is_embargoed(self):
 
     class Meta:
         abstract = True
+
+
+class ACLAlertMixin(ACLMixin, AlertMixin):
```

Comment:
Is it really necessary to couple these two? because if not, I would just implement the validations in `ACLMixin` and let implementors choose whether they want alerts or hard errors in any given model by choosing to inherit from it or not.

Reply:
The whole point of mixins is that you can mix and match as you choose, and in the cases in which you added this `ACLAlertMixin` the `AlertMixin` was already present, so I don't think there's a big benefit to coupling them together

Reply:
You have a valid point. When I chose to make the merging ACLAlertMixin it was based on the fact that it requires functionality of both parent mixins. I was not sure whether this functionality is always needed in the models of the parent mixins. I see that ACLMixin is still left there in the models while the AlertMixin was fully replaced. I thinking whether we might might need a model with alerts which will not have ACLs. Currently it is not a case so I may merge them. On the other hand building the functionality one on another looks more readable to me here.

Reply:
It doesn't really require functionality of the alert mixin, raising an error is perfectly fine, if you want it to treat it as an alert sometimes then you can add `AlertMixin` to the mix, it's the whole goal of composability.

If we do not want to have ACL without alerts, then might as well make `ACLMixin` inherit from `AlertMixin` and use that instead.

`FlawComment` seems to inherit from `ACLMixin` alone, btw, not sure if that's intended.

Reply:
One important point of the `AlertMixin` is that you can explicitly suppress the error raising by `raise_validation_error=False` and it is used a lot in the tests and on the collector side so because of this I think any model with the custom validations should be the `AlertMixin`. If the native Django validations are enough the it can be just a `ValidateMixin` but I see a clear hierarchy there.

To the composability having ACLs is a different feature than having validation or alerts so IMO it is correct to have it defined as an independent mixin. However having validated ACLs is again a composition - of having ACLs and alerts. And as it is quite a typical situation that we require validated ACLs I made it a mixin.

Should I make it as an independent mixin? I think that not as it clearly depends on the functionality of the two. Should I make it a part of one of the existing mixins? I am not sure. It is an extended composition of them. Why would this be against composability? For your model you can choose whatever composition you want except the options violating the dependencies.

Or maybe I do not understand well now :smile_cat: The `FlawComment` is not much being handled. There is no way how to set it so I do not care about it much in the context of two-way sync.

Reply:
@Elkasitu I could imagine if we agree that every ACLs should be valid that we would merge `ACLMixin` and `ACLAlertMixin`. The resulting let us say `ACLMixin` would definitely require `AlertMixin` as a base. Then we would have a clear linear hierarchy. It would be less composable but simpler.

Reply:
For me it doesn't make sense:

With your patch you have to either choose between ACL with no validations (does that even make sense, why would you allow duplicate groups, semantic errors, etc.) or ACL with validation and alerts (what if you don't want alerts?).

You can still achieve the same with something like

```python
class ACLAlertMixin(ACLMixin, AlertMixin):
    pass
```

if you define validations in `ACLMixin` and you really want the shortcut class with both components.

Reply:
I agree that ACLs without any validations makes a little sense. On the other hand I am not sure we should have validations without alerts. The reason is that such validations would always result in an exception so such model eg. could not be loaded by a collector as the some broken data are expected there but the exceptions could not be suppressed. On the other hand it could be implemented your way and all the models fetched by collectors would just use the composite `ACLAlertMixin`. I could do it this way.

Reply:
I reworked the PR so now only the `ACLMixin` is extended and now new mixin is introduced.